### PR TITLE
Prepare for building native images for all storages and tenant-manager

### PR DESCRIPTION
**Pull request by famartinrh**
This is not 100% finished, images compile and look like it all works fine, but I still want to improve some things

* build scripts are not 100% , I did not add the native images build to the default build script because that will increase build times a lot, therefore our release jobs are not handling the native images, we need to think how to address this @riprasad 
* I want to add some comments to explain why some things have been added
* I have to ensure that quarkus tests are executed against the native executables
* I want to run integration tests or k8s tests using the native images
* With the lessons learnt from #1482 we could re-implement the tenant-manager client (but this will wait)
* TODO - add `@RegisterForReflection` to all datamodel generated from Apicurio Studio

### CODE REVIEW

File: app/pom.xml [hunk 1]
```diff
@@ -145,8 +145,8 @@
 
         <!-- Kafka Connect -->
         <dependency>
-            <groupId>org.apache.kafka</groupId>
-            <artifactId>connect-api</artifactId>
+            <groupId>io.quarkus</groupId>
```
Comment (carlesarnal) [inline app/pom.xml:148]: I do not see the relation between this and the PR purpose. Changing from standard apache libraries to quarkus ones for better native compatibility?

**Comment by carlesarnal** [comment]
First review seems mostly ok to me.

**Comment by riprasad** [comment]
> * build scripts are not 100% , I did not add the native images build to the default build script because that will increase build times a lot, therefore our release jobs are not handling the native images, we need to think how to address this @riprasad

Sure! Feel free to create an issue and assign the task to me.

**Review by riprasad — APPROVED** [review]
Workflow changes LGTM

File: app/pom.xml (hunk 1 above)
Comment (EricWittmann) [reply app/pom.xml:148]: Agreed - and isn't Kafka Connect definitely different from the quarkus kafka client?

File: app/pom.xml [hunk 2]
```diff
@@ -252,6 +252,12 @@
             <artifactId>keycloak-admin-client</artifactId>
             <scope>test</scope>
         </dependency>
+        
+        <dependency>
+            <groupId>net.java.dev.jna</groupId>
+            <artifactId>jna</artifactId>
+        </dependency>
```
Comment (EricWittmann) [inline app/pom.xml:259]: Oooo what's this?

File: app/src/main/java/io/apicurio/registry/ccompat/dto/CompatibilityLevelDto.java [hunk 3]
```diff
@@ -41,6 +42,7 @@
 @Getter
 @EqualsAndHashCode
 @ToString
+@RegisterForReflection
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/ccompat/dto/CompatibilityLevelDto.java:45]: Remind me why these annotations are needed?

File: distro/docker/src/main/docker/Dockerfile.native [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
+RUN chown 1001 /work \
+    && chmod "g+rwX" /work \
+    && chown 1001:root /work
+    
+COPY --chown=1001:root target/apicurio-registry-*-runner /work/application
 
 EXPOSE 8080
+USER 1001
 
 CMD ["./application", "-Dquarkus.http.host=0.0.0.0"]
```
Comment (EricWittmann) [inline distro/docker/src/main/docker/Dockerfile.native:14]: No more `JAVA_OPT` - but I assume ENV vars still work...

**Review by EricWittmann — APPROVED** [review]

File: app/pom.xml (hunk 1 above)
Comment (famartinrh) [reply app/pom.xml:148]: it's different yes, but looks like we are using some APIs from kafka clients, and this was required in order to compile and with `connect-json` (just below) we still have the kafka-connect api

File: distro/docker/src/main/docker/Dockerfile.native (hunk 4 above)
Comment (famartinrh) [reply distro/docker/src/main/docker/Dockerfile.native:14]: env vars for everything now

**Comment by famartinrh** [comment]
last discovery, h2 embeddeed database cannot be compiled to native, so for now we will only have native images for sql storage and tenant-manager

File: app/src/main/java/io/apicurio/registry/ccompat/dto/CompatibilityLevelDto.java (hunk 3 above)
Comment (famartinrh) [reply app/src/main/java/io/apicurio/registry/ccompat/dto/CompatibilityLevelDto.java:45]: yes, exactly, there are some classes that quarkus is smart enough to detect and automatically register them, but some others like when we manually SerDe them using `ObjectMapper` not, so we have to add this annotation

Comment (famartinrh) [reply app/src/main/java/io/apicurio/registry/ccompat/dto/CompatibilityLevelDto.java:45]: yes, exactly, there are some classes that quarkus is smart enough to detect and automatically register them, but some others like when we manually SerDe them using `ObjectMapper` not, so we have to add this annotation

**Comment by famartinrh** [comment]
We have a blocker, the http auth permissions are build time properties (but we discovered the properties can be changed at runtime when running in jvm mode) [here some doc](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.auth.permission.-permissions-.enabled)

This makes our auth tests to fail when testing the native images. However looks like we have authentication (because test `testWrongCreds` works fine), but we don't have authorization...

We need to make a decision, IMO it's either make a specific build with the configuration options we need in a PROD deployment or accept the limitations and be forced to offload the authorization to some sidecar technology

@Apicurio/developers thoughts?

**Comment by carlesarnal** [comment]
> We have a blocker, the http auth permissions are build time properties (but we discovered the properties can be changed at runtime when running in jvm mode) [here some doc](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.auth.permission.-permissions-.enabled)
> 
> This makes our auth tests to fail when testing the native images. However looks like we have authentication (because test `testWrongCreds` works fine), but we don't have authorization...
> 
> We need to make a decision, IMO it's either make a specific build with the configuration options we need in a PROD deployment or accept the limitations and be forced to offload the authorization to some sidecar technology
> 
> @Apicurio/developers thoughts?

Adding a comment so you can have all the information. I just learned today that being able to enable/disable the HTTP policies at runtime in JVM is a bug that will be removed in future Quarkus versions. For me, it's preferable to have a specific build with the production config and use a reduced of the testsuite to test it. I've asked if they're are planning to make this property a runtime property but that that's not the case.

As a second possibility, we could try to override quarkus behaviour with our own processors as I did with the `CustomAuthenticationMechanism` for the basic authentication. That way we can enable the property by default but process HTTP permission policies on our own at runtime. Basically check if authz is enabled (environment property), allow if not and delegate to Quarkus when true.

**Comment by famartinrh** [comment]
I think the approach from carles about overriding quarkus behavior could be a good idea, at least worth a try.

I vote for merging the work done in this PR, it won't hurt, but commenting the new GH workflow. Later we can experiment with the carles idea and test it with the commented GH workflow

ok? @Apicurio/developers

**Comment by EricWittmann** [comment]
I am in favor of creating our own (optional) role checking, possibly via interceptor.  Or possibly via the new @famartinrh storage delegate pattern (might be too late to do it here).  I actually never really liked the configuration approach used in Quarkus.  Config is fine, but when the config is build-time config, I don't see the difference between that and code.  Assuming we can get access to the Auth Token and check it for the existence of required roles, I don't see why we couldn't do that easily.

**Pull request merged**